### ISSUE TITLE
Feature Request: Set Options Via Environment or .migraterc File

### ISSUE THREAD

**Issue by zebulonj** [tj/node-migrate#101]
There may be a better way to accommodate this, so let me frame it in terms of a use-case...

I'm storing state in a database (using the `--store` option to designate) have migrations in a non-standard location (`--migrations-dir=...`) and have written my migrations in ES6 (`--compiler=...`), but it's a pain to specify all those command-line options every time I run migrate[1].

It'd be nice to be able to specify those options in a per-repository `.migraterc` file or via environment (managed by other means) on each repository.

[1] I initially solved this by adding a `migrate:up` script to my `package.json`, but as my needs become more complex that isn't cutting it. I've run into problems using `package.json` scripts to handle rolling pack individual migrations, etc.

**Comment by zebulonj** [comment]
P.S. Happy to submit a pull-request to accomplish this, but wanted some general buy-in (or course correction) before I invest any time.

**Comment by LinusU** [comment]
We already have support for reading environmental variables from the file `.env` if it exists. That being said, I think that these kinds of options shouldn't go there because they aren't "environment" dependent, but rather dependent on the setup of the project.

I think that a `.migraterc` file sounds very good! Potentially we could also leverage some `package.json` key under `config`, or potentially support both...

Basically, I think it's a great idea!

**Comment by zebulonj** [comment]
@LinusU I looked at the `.env` support, and agree that this is a different use case. Since it sounds like this isn't a bad idea out of the gate, I'll get the broad strokes into a pull-request and throw it back to you for feedback.

**Comment by zebulonj** [comment]
@LinusU Here's a question for you: should this file be a `JSON` file (similar to `.babelrc`) or a JavaScript file (similar to `webpack.config.js`). The advantage of the later being that logic could be included to... say... resolve a path name or configure a migration state store.

**Comment by wesleytodd** [comment]
Something similar to this has come up multiple times now, so I think it has merit for sure.  I think the `rc` file idea is good.  Maybe take a look at using [`rc`](https://www.npmjs.com/package/rc)?

**Comment by zebulonj** [comment]
Great. I have something rough working locally. Expect a pull request later today. I would still be interested in thoughts on my last question (although your reference to `rc` does offer an implicit answer).

**Comment by wesleytodd** [comment]
I believe that it is best for a module like this one to remain as agnostic and simple as possible. 

I don't think coming up with our own semantics around file look up is a good idea, so that is why I suggested `rc`.  I believe that `rc` allows for an assortment of file formats, but does not use `require` to pull in the files, so having javascript logic is not really possible.

IMO having something like a `webpack.config.js` is just too complicated for the needs of this package.  And a simple format like `ini` or `json` should serve everyone's purposes.

**Comment by LinusU** [comment]
Hmm, at first glance I really liked `rc` but thinking more about it I'm not so sure. It seems to be very focused on getting the information from a number of different places, most of them being _environment depdendent_.

To me it feels like the settings we are talking about here is highly _project specific_, and thus should only be able to be configured from within the project.

The problem with being able to have a (e.g.) `$HOME/.migraterc` file is that we will see projects which depend on values there, and then won't work when cloned into a new computer.

I personally would prefer if the only way to set this things would be via a `.migraterc`. Using `rc` would mean that they can be set via environmental variables and global files, which I would consider a huge negative...

Just my 2¢ 🙂

**Comment by zebulonj** [comment]
I would tend to agree, particularly on the point that these are project specific settings. That said, using `rc` does get us standard format support (both JSON and .ini) out of the box.

I defer to the maintainers.

**Comment by LinusU** [comment]
> (both JSON and .ini)

Personally not a fan of supporting multiple formats either 😄 

...and there isn't really a good spec for `ini` files so we will probably look ourselves to that engine. I would rather prefer TOML instead, which is very similar but has a very strict spec for how the language works.

**Comment by zebulonj** [comment]
TOML looks nice (honestly, never heard of it before), but if we're going to keep it simple can we just stick to JSON, which is what we see in just about any other similar tooling on this stack (prime example: `.babelrc`)?

There are two areas where I'd like clear direction before I invest any additional effort in this:
1. Where can configuration be set (e.g., local `.migraterc` file, environment variables, user $HOME folder,...).
2. What formats are to be supported: JSON, INI, TOML,...

I think @LinusU makes a reasonable point about these settings being project specific:
> To me it feels like the settings we are talking about here is highly project specific, and thus should only be able to be configured from within the project.

So too does @wesleytodd:
> I don't think coming up with our own semantics around file look up is a good idea...

My $0.02 on where these balance out: use `rc` because it implements a fairly widely understood paradigm around configuration. It is up to the consumer how she or he uses the power it endows. And... if this is going to be a sticking point... support the smallest common denominator (which seems to be JSON in a project-local `.migraterc`).

Let's aim for where there is consensus, as a starting point. Can always expand outward from there.

**Comment by wesleytodd** [comment]
I think you both  have great points.  I don't really care much for `rc` in particular, it was just a highly used module, so I mentioned it.

As for the two questions:

> 1. Where can configuration be set

I agree that in 98% of cases it will be project specific, but I think it is also reasonable for having shared profiles with would be nicely supported with a `$HOME/.migraterc` or a `~/project/.migraterc`.  For example, on  my work computer I might want all my projects to use a work specific migration template.  Which I could then override on a project by project basis.  So that is one argument for the global config option.  Not a requirement, but one situation I could think of.

> 2. What formats are to be supported

I am happy just doing json.  But I also think remaining agnostic so that @LinusU could use TOML and someone who loves ini (or yaml) could use ini (or yaml) would be best.  I am pretty sure `rc` is not that agnostic, which is disappointing.

**Comment by wesleytodd** [comment]
Interesting, so I am at work and didn't really read through what `rc` does until now.  So I don't like the idea of having two command line flag parser runs.  There was another issue with `commander`, and I was actually looking at `minimist` (which is what `rc` is using under the hood).  Unfortunately there are some unresolved issues, I emailed and tweeted @substack about helping out on it but he never responded.

All that being said, I don't mind the idea of switching fully to use `rc` and remove `commander`.  I actually just used `minimist` on another cli I wrote (https://github.com/wesleytodd/shush-cli), and it worked well.

Anyway, now that I have had time to actually look at it, I think #102 is not a good idea for now because of this issue.  If substack responds and grants me access to minimist, then I would like to switch to it, at which point we could also just switch to using `rc`.  But if he doens't the issues and lack of support seem like it is not a great choice.

**Comment by zebulonj** [comment]
@wesleytodd Let’s say `migrate` moves to `minimist` (which I’ve used on other projects too), isn’t there still the issue of redundant processing of the command-line flags?

Can you point me at the issues you’ve had with `minimist`?

Here’s how I’d propose moving forward:

1. Let me add support for a JSON formatted `.migraterc` in the project root (`process.cwd()`). I think there’s consensus about both that format (JSON) and that source (`$PROJECT/.migraterc`).

2. If we reach consensus about something broader (e.g. `rc`), we expand options then. 

I don’t see any risk that iterating in this way would introduce breaking changes down the line, and perfect doesn’t become the enemy of good.

**Comment by LinusU** [comment]
> Here’s how I’d propose moving forward: [...]

This is exactly how I think it should be done 👍

**Comment by wesleytodd** [comment]
Ok, I am on board with moving forward with this plan.

> isn’t there still the issue of redundant processing of the command-line flags?

I was meaning we would replace all of `commander` with `rc`, so it was back to only one step (just using `minimist` under the hood).  But for now I think your plan is best.

> Can you point me at the issues you’ve had with minimist?

Mainly there are a bunch of pending PR's and issue reports.  Only one of which I had hit.  But if there are not active maintainers, I do not want to move this library to it.  And I offered to @substack to take over maintenance, but heven't even heard back on that offer.

**Comment by wesleytodd** [comment]
This should be resolved with https://github.com/tj/node-migrate/pull/88, right?
